src/lsm/superblock.zig

```diff
@@ -0,0 +1,1491 @@
+            if (a.sector.replica != b.sector.replica) return error.ReplicaDifferent;
+        }
+
+        // Even the best copy with the most quorum still has inadequate quorum.
+        if (!b.valid) return error.QuorumLost;
+
+        // Verify that the parent copy exists:
+        for (quorums.slice()[1..]) |a| {
+            if (a.sector.checksum == b.sector.parent) {
+                assert(a.sector.checksum != b.sector.checksum);
+
+                if (!a.valid) {
+                    return error.ParentQuorumLost;
+                } else if (a.sector.sequence >= b.sector.sequence) {
+                    return error.SequenceNotMonotonic;
+                } else if (a.sector.sequence % 2 == b.sector.sequence % 2) {
+                    // The parent must reside in the alternate copy to guarantee that we are able to
+                    // detect when the working quorum is lost.
+                    return error.SequenceNotMonotonic;
+                } else if (!a.sector.vsr_state.monotonic(b.sector.vsr_state)) {
+                    return error.VSRStateNotMonotonic;
+                } else {
+                    assert(b.sector.magic == .superblock);
+                    assert(b.sector.valid_checksum());
+
+                    return b.sector;
+                }
+            }
+        }
+
+        return error.ParentNotFound;
+    }
+
+    fn count_copy(
+        quorums: *Quorums,
+        copy: *const SuperBlockSector,
+        index: usize,
+        threshold: u8,
+    ) void {
+        assert(index < superblock_copies_max);
+        assert(threshold >= 2 and threshold <= 5);
+
+        if (!copy.valid_checksum()) {
+            log.debug("copy: {}/{}: invalid checksum", .{ index, superblock_copies_max });
+            return;
+        }
+
+        if (copy.magic != .superblock) {
+            log.debug("copy: {}/{}: not a superblock", .{ index, superblock_copies_max });
+            return;
+        }
+
+        if (copy.copy == index) {
+            log.debug("copy: {}/{}: checksum={x} sequence={}", .{
+                index,
+                superblock_copies_max,
+                copy.checksum,
+                copy.sequence,
+            });
+        } else {
+            // If our read was misdirected, we definitely still want to count the copy.
+            // We must just be careful to count it idempotently.
+            log.err(
+                "copy: {}/{}: checksum={x} sequence={} misdirected from copy={}",
+                .{
+                    index,
+                    superblock_copies_max,
+                    copy.checksum,
+                    copy.sequence,
+                    copy.copy,
+                },
+            );
+        }
+
+        var quorum = quorums.find_or_insert_quorum_for_copy(copy);
+        assert(quorum.sector.checksum == copy.checksum);
+        assert(quorum.sector.equal(copy));
+
+        quorum.count.set(copy.copy);
```

Comment:
> It should discard (with just a log) any wrong-cluster superblocks.

Yes, a fail-fast approach is definitely what we would do in another context, but see my previous comment. At this point, we only want to open the data file and work out what it's telling us about the data file. For example, you can imagine this being used also in a script utility, something like smartctl where you're just inspecting the disk stats.

> Otherwise this line could panic if the bitset is too small — that is, the other cluster (whose superblock was mistakenly read) might be configured to store more copies than this cluster.

The superblock layout should always be "well-known" across all TigerBeetle clusters. Tweaking `superblock_copies` is not recommended, as it's more part of our "generic distributed DB framework" approach to how we're building TigerBeetle. The constant itself will move into `config_comptime.zig`. It's the kind of constant one might change only if one were implementing a whole different state machine on top of the framework, building a whole other database reusing our "global consensus and local storage" framework.

Reply:
Instead of verifying cluster here, what about checking `version` against `SuperBlockVersion`?